### PR TITLE
Fabric: Refactor dir and theme creation

### PR DESCRIPTION
**Pull request by micahgodbolt**
#### Pull request checklist

We were adding LTR to Fabric controls that didn't need it, and missing some edge cases. This logic is better laid out and documented. 

#### Description of changes

(give an overview)

#### Focus areas to test

(optional)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/11808)

### CODE REVIEW

File: packages/office-ui-fabric-react/src/components/Fabric/Fabric.base.tsx [hunk 1]
```diff
@@ -47,13 +49,21 @@ export class FabricBase extends React.Component<
     const { as: Root = 'div', theme, dir } = this.props;
     const classNames = this._getClassNames();
     const divProps = getNativeProps<React.HTMLAttributes<HTMLDivElement>>(this.props, divProperties, ['dir']);
-    const componentDir = getDir(theme, dir);
-    const parentDir = getDir(theme);
+    const getDir = getAllDir(theme, dir);
+    // If Fabric dir !== theme.RTL
+    // Or If theme.RTL !== page RTL
+    // Then we need to set dir on the Fabric component
+    const setComponentDir = getDir.component !== getDir.parent || getDir.component !== getDir.page;
```
Comment (aneeshack4) [inline packages/office-ui-fabric-react/src/components/Fabric/Fabric.base.tsx]: if getDir is being replaced by getAllDir above how are these calls to getDir still valid?

Comment (micahgodbolt) [reply packages/office-ui-fabric-react/src/components/Fabric/Fabric.base.tsx]: i just reorganized the logic.

File: packages/office-ui-fabric-react/src/components/Fabric/Fabric.base.tsx [hunk 2]
```diff
@@ -18,14 +18,16 @@ import { ITheme, createTheme } from '../../Styling';
...
-const getDir = memoizeFunction((theme?: ITheme, dir?: IFabricProps['dir']) => {
-  if (dir) {
-    return dir;
-  }
-  if (theme && theme.rtl !== undefined) {
-    return theme.rtl ? 'rtl' : 'ltr';
-  }
-  return getRTL() ? 'rtl' : 'ltr';
+
+const getAllDir = memoizeFunction((theme?: ITheme, dir?: IFabricProps['dir']) => {
```
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/Fabric/Fabric.base.tsx]: `getAllDir`  - can you renaming this method to be more meaningful? or leave it as `getDir`?

File: packages/office-ui-fabric-react/src/components/Fabric/Fabric.base.tsx [hunk 3]
```diff
@@ -18,14 +18,16 @@ import { ITheme, createTheme } from '../../Styling';
...
+  const themeRTL = getRTL(theme) ? 'rtl' : 'ltr';
```
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/Fabric/Fabric.base.tsx]: rename to `themeDir`

File: packages/office-ui-fabric-react/src/components/Fabric/Fabric.base.tsx [hunk 4]
```diff
@@ -18,14 +18,16 @@ import { ITheme, createTheme } from '../../Styling';
...
+  const pageRTL = getRTL() ? 'rtl' : 'ltr';
```
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/Fabric/Fabric.base.tsx]: rename to `pageDir`

File: packages/office-ui-fabric-react/src/components/Fabric/Fabric.base.tsx (hunk 2 above)
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/Fabric/Fabric.base.tsx]: why are we memoizing the function?

File: packages/office-ui-fabric-react/src/components/Fabric/Fabric.base.tsx (hunk 1 above)
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/Fabric/Fabric.base.tsx]: I'm super confused here. Some of the dir logic is in the `getAllDir` method, and some is here.

Can you just have a single method that gets the correct `dir` value and assigns it to the `dir` attribute?

```tsx
<Root dir={ getDir(theme, props) } ... />
```

**Comment by micahgodbolt** [comment]
oh, just saw these comments. I'll look at them today.

File: packages/office-ui-fabric-react/src/components/Fabric/Fabric.base.tsx (hunk 2 above)
Comment (micahgodbolt) [reply packages/office-ui-fabric-react/src/components/Fabric/Fabric.base.tsx]: it was memoized when i got there :D

Comment (micahgodbolt) [reply packages/office-ui-fabric-react/src/components/Fabric/Fabric.base.tsx]: Original call was memoized, so I copied that pattern.

https://github.com/OfficeDev/office-ui-fabric-react/blob/master/packages/office-ui-fabric-react/src/components/Fabric/Fabric.base.tsx#L20

Comment (micahgodbolt) [reply packages/office-ui-fabric-react/src/components/Fabric/Fabric.base.tsx]: but i can see how a large theme is quite different than this call..will remove

**Comment by micahgodbolt** [comment]
@dzearing what do you think? is this worth trying to pull in, or leave it as is?

File: packages/office-ui-fabric-react/src/components/Fabric/Fabric.base.tsx [hunk 5]
```diff
@@ -18,15 +18,21 @@ import { ITheme, createTheme } from '../../Styling';
...
+  const pageDir = getRTL() ? 'rtl' : 'ltr';
+  const componentDir = dir ? dir : contextDir;
+  return {
+    // If Fabric dir !== contextDir
+    // Or If contextDir !== pageDir
+    // Then we need to set dir of the Fabric root
+    rootDir: componentDir !== contextDir || componentDir !== pageDir ? componentDir : undefined,
+    // If dir !== contextDir || pageDir
+    // then set contextual theme around content
+    needsTheme: componentDir !== contextDir
```
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/Fabric/Fabric.base.tsx:33]: needsDir?

**Comment by micahgodbolt** [comment]
OMG...i just saved myself a ton of headache by using the in-github merge conflict editor.

**Comment by micahgodbolt** [comment]
All passing and signed off!
